### ISSUE TITLE
DeOldify vs SuperRes-GAN Workbook

### ISSUE THREAD

**Issue by AshleyRudland** [jantic/DeOldify#262]
This is not an issue. An educational question. Thank you for DeOldify, learnt a lot from reading the source and running the training.

I've played with FastAI SuperRes-GAN workbook 100 times for removing artifacts from images. I've also tried your DeOldify training workbook. From the outside the the training looks very similar but under the hood DeOldify is written from scratch.

I completely understand these are very different problems and DeOldify is about colorising black & white photos. But I'm trying to understand the differences in the UNET architecture and why you wrote your own instead of using what's in Fast AI.

Forgive the really dumb question - but if I can understand better, that will allow me to apply unet to new problems and will allow me to think about how to architect a unet.

Thanks in advance Jason!

**Comment by jantic** [comment]
@AshleyRudland  Hey, great question actually! So there are a few things going on here:

1. I needed to be able to add both spectral norm and batch norm for each layer.  FastAI's DynamicUnet doesn't support this, and it's insufficient to omit one or the other.  Believe me, I've tried.
2. I needed to be able to customize the width (number of channels per layer) with a multiplier (via nf_factor). Widening the models proved to make a big difference.
3. DynamicUnetDeep is a slight customization based on what has already been spelled out in #1 and #2.  But DynamicUnetWide is significantly different in the architecture in that it cuts the number of layers in favor of having fewer but wider layers. This proved to make video more stable.

My original intent was to eventually do a pull request on these customizations for the core FastAI library,  but naturally I let that slide and forgot about it.

I will say this too as a big hint on stuff I haven't released (it's closed source): It's worthwhile to reexamine the Unet for memory efficiency and effectiveness. There's a lot of low hanging fruit there.
